File: matlab/examples/ex_simple_image.m
% This example shows how to call FovVideoVDP and visualize its results. 

if ~exist( 'fovvdp', 'file' )
    addpath( fullfile( pwd, '..') );
end

I_ref = imread( '../../example_media/wavy_facade.png' );
%I_ref = imread( 'tree.jpg' );

% Generate distorted images as needed.
% We store distorted images so that we can compare Matlab/Python results
noise_fname = '../../example_media/wavy_facade_noise.png';
if isfile( noise_fname )
    I_test_noise = imread( noise_fname );
else
    I_test_noise = imnoise( I_ref, 'gaussian', 0, 0.003 );
    imwrite( I_test_noise, noise_fname );
end

blur_fname = '../../example_media/wavy_facade_blur.png';
if isfile( blur_fname )
    I_test_blur = imread( blur_fname );
else
    I_test_blur = imgaussfilt( I_ref, 2 );
    imwrite( I_test_blur, blur_fname );
end

[Q_JOD_noise, diff_map_noise] = fvvdp( I_test_noise, I_ref, 'display_name', 'standard_4k', 'heatmap', 'threshold' );
[Q_JOD_blur, diff_map_blur] = fvvdp( I_test_blur, I_ref, 'display_name', 'standard_4k', 'heatmap', 'threshold' );

fprintf( 1, '=== Noise Q_JOD = %g\n', Q_JOD_noise);
fprintf( 1, '=== Blur Q_JOD = %g\n', Q_JOD_blur);


clf
subplot( 2, 2, 1 );
imshow( I_test_noise );
title( 'Test image with noise' );

subplot( 2, 2, 3 );
imshow( diff_map_noise );
title( sprintf('Noise; Quality: %.3f JOD\n', Q_JOD_noise ) );


subplot( 2, 2, 2 );
imshow( I_test_blur );
title( 'Test image with blur' );

subplot( 2, 2, 4 );
imshow( diff_map_blur );
title( sprintf('Blur; Quality: %.3f JOD\n', Q_JOD_blur ) );

File: pytorch_examples/ex_simple_image.py
import os
import numpy as np
import matplotlib.pyplot as plt
import torch
import ex_utils as utils
import pyfvvdp

debug = False


I_ref = pyfvvdp.load_image_as_array(os.path.join('example_media', 'wavy_facade.png'))

noise_fname = os.path.join('example_media', 'wavy_facade_noise.png')
if os.path.isfile(noise_fname) and debug:
    I_test_noise = pyfvvdp.load_image_as_array( noise_fname )
else:
    std = np.sqrt(0.003)
    I_test_noise = utils.imnoise(I_ref, std)

blur_fname = os.path.join('example_media', 'wavy_facade_blur.png')
if os.path.isfile(blur_fname) and debug:
    I_test_blur = pyfvvdp.load_image_as_array( blur_fname )
else:
    sigma = 2
    I_test_blur = utils.imgaussblur(I_ref, sigma)

fv = pyfvvdp.fvvdp(display_name='standard_4k', heatmap='threshold')

# predict() method can handle numpy ndarrays or PyTorch tensors. The data
# type should be float32, int16 or uint8.
# Channels can be in any order, but the order must be specified as a dim_order parameter. 
# Here the dimensions are (Height,Width,Colour)
Q_JOD_noise, stats_noise = fv.predict( I_test_noise, I_ref, dim_order="HWC" )
noise_str = f'Noise - Quality: {Q_JOD_noise:.3f} JOD'
print( noise_str )

Q_JOD_blur, stats_blur = fv.predict( I_test_blur, I_ref, dim_order="HWC" )
blur_str = f'Blur - Quality: {Q_JOD_blur:.3f} JOD'
print( blur_str )

f, axs = plt.subplots(2, 2)
axs[0][0].imshow( I_test_noise/(2**16 - 1) )
axs[0][0].set_title('Test image with noise')
axs[0][0].set_xticks([])
axs[0][0].set_yticks([])
axs[0][1].imshow( I_test_blur/(2**16 - 1) )
axs[0][1].set_title('Test image with blur')
axs[0][1].set_xticks([])
axs[0][1].set_yticks([])
axs[1][0].imshow( stats_noise['heatmap'][0,:,0,:,:].permute([1,2,0]).to(torch.float32).numpy() )
axs[1][0].set_xticks([])
axs[1][0].set_yticks([])
axs[1][0].set_title(noise_str)
axs[1][1].imshow( stats_blur['heatmap'][0,:,0,:,:].permute([1,2,0]).to(torch.float32).numpy() )
axs[1][1].set_xticks([])
axs[1][1].set_yticks([])
axs[1][1].set_title(blur_str)

f.show()
plt.waitforbuttonpress()
